### ISSUE TITLE
[monitor-workbook] IoT Edge device details - Messaging - If module doesn't ACK messages it seems that corresponding queue size increases out of thin air

### ISSUE THREAD

**Issue by alaendle** [Azure/iotedge#5662]
In a situation where the receiving module doesn't acknowledge messages it seems that the corresponding queue increases out of thin air. It looks like the sending module doesn't transmit any message - but that is just wrong. Please take a look at the screenshot and let me know if I could provide further insights.

![2021-10-11_13h36_46](https://user-images.githubusercontent.com/969523/136785429-e5f7d718-bff4-408d-9f6e-300568a6dfa0.png)

If the receiving module (here "visu-poc-influxdb-ingest::ingest") couldn't process incoming messages (here from "visu-poc-rabbitmq-gateway::unide_process") the corresponding queue increases (what is correct) (here from 131k to 179k messages), but the workbook couldn't explain where the messages coming from - it seems that the sender that produces the messages doesn't send any message at all (here count equals 0 from "visu-poc-rabbitmq-gateway::unide_process").

@veyalla  - maybe this is something for you?

**Comment by micahl** [comment]
@alaendle if you select the first child below the one you currently have selected (i.e. visu-poc-rabbitmq-gateway::unide_process) it will show a timeline of messages sent from visu-poc-rabbitmq-gateway::unide_process to the visu-poc-influxdb-ingest::ingest.  Does it show the counts of messages sent (which weren't processed)?

**Comment by alaendle** [comment]
![2021-10-19_08h26_46](https://user-images.githubusercontent.com/969523/137856000-d852744c-02b1-47d0-a6a3-aac1788f6e2f.png)

**Comment by nyanzebra** [comment]
@micahl & @alaendle, not sure if the fix I was working on will be applicable here, but I can try testing similar test and see what results I get

**Comment by nyanzebra** [comment]
@alaendle, I am looking to reproduce this situation and have a few more questions just to make sure I am replicating the exact same scenario.

The situation is that you have a sender and a receiver of messages (we will call these ModuleSend and ModuleRecv). Using QoS1? ModuleSend will send messages to ModuleRecv. But because ModuleRecv is not ack'ing the messages there are effectively 0 new messages to report on. 

So want to confirm the flow you are testing is:
```
ModuleSend -> Broker -> ModuleRecv -> Broker -> $Upstream
```
or
```
ModuleSend->Broker->ModuleRecv
                             |
                      $Upstream
```

Also, were there any QoS1 or QoS2 messages that were sent before and then you deployed updated modules? Just want to rule out that the queue wasn't built up from before, because we do store and forward and when the broker comes up and sees unprocessed messages it will say that is the queue size (because those messages were never processed).

**Comment by alaendle** [comment]
@nyanzebra, it is even simpler; the flow is just
```
  ModuleSend->EdgeHub->ModuleRecv
```
The modules are simply connected via a route like
```
FROM /messages/modules/ModuleSend/outputs/* INTO BrokeredEndpoint(\"/modules/ModuleRecv/inputs/input\")
```

Both modules make use of the Azure IoT C SDK (latest LTS release), use QoS1 and MQTT - no module updates happend.

I haven't tried to reproduce the behaviour with **v1.2.5** or with a _new/clean_ installation, but I still hope it's easy enough for you to reproduce. My last experience was that it is sufficient to just stop the receiving module to experience the strange numbers in the dashboard. If you need additional information just let me know 😃

**Comment by nyanzebra** [comment]
Okay, perfect, let me try with 1.2.4 and 1.2.5 and get back to you :)

**Comment by nyanzebra** [comment]
What I did:
Started with fresh install and deployed (1.2.4) with SimulatedTemperatureSensor sending messages to a custom C Module through broker which in turn sent to upstream through broker. I am not seeing similar results...

![image](https://user-images.githubusercontent.com/6096983/143953285-4bb6c687-41d3-4a87-9e56-3ecb6894f43e.png)
![image](https://user-images.githubusercontent.com/6096983/143953342-2b666af2-4794-47a7-b83d-1b7ebadb2cf5.png)

As can be seen here, my SimulatedTemperatureSensor module is still showing positive messages being sent. Was there anything required on sender side that needs to be done? I have similar route, similar setup, but am not seeing this situation occur.

I stopped C Module by writing a quick bash script to continuously stop the docker container. In case this is still easily reproducible, could you try looking at the logs of sender (if any) and see if it is actually sending messages? I am curious if this is a problem with representation of the data.

**Comment by alaendle** [comment]
Hi @nyanzebra, I think you already reproduced the problem. Just focus on the marked timeframe. Now ask yourself, why is the queue filling up? In the graph below the one showing the queue length it seems that no messages are sent to `CModuleSend::input`. O.k. it might be the case that this graph only shows the messages that really got delivered from the edgeHub to the `CModuleSend::input` - than zero is correct. But now the question remains - why is the queue filling up? And I suspect that if you click on `SimulatedTemperatureSensor::temperatureOutput` (unfortunately this screenshot is missing) you wouldn't spot any message flow from this output to the queue. Long story short, by just looking at the dashboard you couldn't tell why the queue is growing.

![143953342-2b666af2-4794-47a7-b83d-1b7ebadb2cf5](https://user-images.githubusercontent.com/969523/143988941-d8f09369-1168-49d1-bcdb-bcf809076483.png)

Hope this helps. Many thanks for your effort - and don't hesitate to contact me if I could contribute further information or if you gain new insights.

P.S.: Also things got more complicated in real life, where you  might just have a slow consumer and not complete offline modules - or you have multiple senders.

**Comment by nyanzebra** [comment]
Ahh, I see, so not just the count of 0 on pic from above is the problem... I added a pic for the SimulatedTemperatureSensor which does show 0 messages being sent.  At least in my case the queue is filling up because the SimulatedTemperatureSensor is still sending messages with a flow like this (should be same for your case or at least close enough):

SimulatedTemperatureSensore -> Broker -> CModule -> Broker -> $upstream.

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/6096983/144023500-05344ba5-e364-4d9a-aa92-242a279df3c4.png">

Let me look through our code and how the queries work on portal for the workbook, I suspect you might be right that it shows only after message is delivered to target and that we either publish the counter later than we maybe should (message sent) or the query is doing something strange. I will get back to you soon on what is happening, as looking at this again from my test, I agree it looks weird...

**Comment by nyanzebra** [comment]
@veyalla looking into our code, I think the problem expressed above is likely due to not having another graph present in the device details. We have two metrics edgehub_messages_sent_total, which is used for the graph where we see 0, and we have edgehub_messages_received_total, something I think that might need to be added. 

When a module publishes a message, it will go to the broker on the device. Only when the message is forward to the destination is the send metric incremented, which makes perfect sense. However, when the broker received messages the receive metrics gets incremented. So, right now it is hard to see why the queue is growing, which should be mapped to the messages received by the broker. This is also different from the queue length, in that the aggregate of receive, sent, and queue length should clue in a user into the rate at which messages are being processed.

**Comment by micahl** [comment]
@alaendle, I have a pre-release change to use the _edgehub_messages_received_total_ metric in the 'Messages sent to ...' visualization.  I believe it will address the issue based on @nyanzebra's comment.  Can you give it a try and let us know? 
 Just include `this flag` in the portal's URL:
[https://portal.azure.com/`?feature.includePreviewTemplates=true&feature.workbookGalleryBranch=users%2Fmicahl%2Fedge-fleetview`#home](https://portal.azure.com/?feature.includePreviewTemplates=true&feature.workbookGalleryBranch=users%2Fmicahl%2Fedge-fleetview#home)

**Comment by alaendle** [comment]
@micahl - Thanks for the preview, looks realistic for the receiver side now - but on my first impression still reports wrong numbers for the sender. 
![grafik](https://user-images.githubusercontent.com/969523/144898040-5a2fecd3-2d4a-40f2-8908-4fb238b5333c.png)
So the module receives around 8k messages - but only 4 k are sent:
![grafik](https://user-images.githubusercontent.com/969523/144898157-8c4220be-9d9b-4c5c-9069-a5271ddb7956.png)
Also I'm missing a value/chart that represents the actual throughput - from hub (queue) to module. This would be helpful to easily detect (in the case the queue grows) if the throughput of the receiving module dropped or if just more messages are send.

**Comment by nyanzebra** [comment]
So it should be the case that messages_sent are the messages from broker to module and messages_received are from module to broker. So, to be clear, from the pictures above with sent at 4k and receive at 8k, the expectation is that both should be 8k? As it should be that if broker is not processing messages fast enough there should be a queue increase and it would be due to the fact that it is receiving more messages than being sent. So, is the problem you are seeing not making sense based on metrics or knowledge that you know should show things both being 8k or is it that we have confusing metric names?

I am not sure I understand the second point on throughput, do you mean from upstream (hub) to downstream(broker->module)?

**Comment by micahl** [comment]
hmm... @nyanzebra from those images it looks like ~25.9K messages in total have been sent (at a rate of 4K per period), but the receiving queue has reported that it has received ~53.6K messages in total (at a rate of about 8K per period).  I'll check the new query using the messages_received to make sure it isn't double-counting.

**Comment by alaendle** [comment]
@nyanzebra, @micahl - please excuse that my last comment wasn't clear enough (it was late in the evening for me and I just wanted to share my latest findings). However, let me add some new context: The screenshots just show a situation where both modules are running without any errors, just the sender sends more messages than the receiver is able to process - therefore the queue piles up. So I wonder about the numbers...
![grafik](https://user-images.githubusercontent.com/969523/144966543-979d03db-7b37-4f7c-b523-1b84e016275f.png)
![grafik](https://user-images.githubusercontent.com/969523/144966998-13a10f79-c191-41aa-849e-e577cfee5472.png)
Please note that these are just my assumption/my interpretation of the graphs (they might be wrong)!

**Comment by micahl** [comment]
@alaendle thanks for clarifying.  The new query was wrong due to a bad assumption on my part.  I've reverted it to the prior query showing messages EdgeHub successfully sent from sender to received.  And I'm now including an additional graph to just show the messages EdgeHub received for the sender's queue.  May still need some refinement, but I've updated the changes if you care to take a look.

**Comment by nyanzebra** [comment]
@alaendle, thanks for the extra info! This clears up what the issue is much more. I think @micahl is out of office for holiday now/soon, so I can try to look into making any other changes if the recent work from @micahl needs any more tweaking for clarity from user perspective. Thanks @alaendle for working with us on making this better!

**Comment by alaendle** [comment]
@micahl @nyanzebra Alright, guess we have a common understanding now. 

Thanks for the latest changes, especially the renaming from "received" to "delivered" makes things now clearer. 

I still believe it would be nice, to reflect the changes in the queue length somehow in the count and throughput numbers of the sender/receiver. 

In conjunction with the third graph for senders I now have all the information at hand to analyze our scenarios - so many thanks for that. However it would be nice if this third graph could somehow depict if the sent messages are live or fed by the queue (e.g. by a different color) - maybe not in the sense of the single message (would violate the fifo character of the queue), but in a way that enables us to easily recognize the amount of messages that drain the backlog. So maybe it is even possible to combine the 2nd and 3rd graph and show send and received values as overlapping charts. But all this is "just" usability/ux-stuff - as said before - with latest changes the numbers a represented, which are needed for our daily work. Thanks again for your support ❤️ .

![grafik](https://user-images.githubusercontent.com/969523/145192018-e39a3775-b430-4107-9527-9824707066e6.png)

Maybe it could be easier to spot if backlog is filled or drained.
![grafik](https://user-images.githubusercontent.com/969523/145192200-45951602-7d30-4746-abe6-cf091b44d434.png)

**Comment by nyanzebra** [comment]
@alaendle so it sounds like the last thing for correctness purposes is to look at the count/rate and verify it is correct. Then we need to address the UX to make things easier to interpret and look at. This sound correct? I will look into the query and see if there is something missing/wrong there.

**Comment by nyanzebra** [comment]
Just a quick update, I have looked through the query for rate and I do think it is wrong. I will update here once there is a test branch to use with an updated query

**Comment by nyanzebra** [comment]
@alaendle & @micahl 

Just want to go through how the graph works and how the metrics are emitted and then see what feedback there is on how the data could be interpreted/shown better.

1. The graph creates a set of aggregates of data for particular routes with child data under each. For instance in the 'curve-grabber' example you showed, the top row is the aggregate of all children and the one below is an individual value. So, in this case the values should be the same.
2. The way the metrics are handled is that they are all in respect to what EdgeHub sees. So, received messages, is actually messages sent from some module to EdgeHub, and send is the inverse.
3. Currently rate is for messages sent from EdgeHub which, while useful info, might be less than what is desired?
4. We currently have all the values in respect to EdgeHub and this also might be part of why things are confusing...

What data and presentation would seem most useful for data interpretation? Are there other metrics you think belong here? Want to get this information so that we might create work items to adjust either code for more metrics or data representation to make things clearer. From internal discussion there are some ideas we have on changing things to improve data representation, but don't want to make any steps before getting more feedback on what *should* be expected to be shown :)

**Comment by jlian** [comment]
@alaendle Any thoughts on @nyanzebra's comment above?

**Comment by alaendle** [comment]
Sorry for the late response; I had days off. Thanks @nyanzebra for the details. What is really key, is the possibility to investigate problems that result in increasing queues. What has changed? Inbound message frequency; outbound latency; etc.. Since most of "my" problems were related to the distinction between inbound and outbound traffic, I think it need to be crystal clear what each number corresponds to (like it was done with the renaming from "received" to "delivered"). 

Basically all this was already stated in https://github.com/Azure/iotedge/issues/5662#issuecomment-988688754.

Regarding point 1 and 3 of https://github.com/Azure/iotedge/issues/5662#issuecomment-1005165507. If I understood it correctly these values (Count; Rate) correspond to the messages send from edgeHub to the module. My expectation was that the at least the line of `smbOutput` shows the messages received by the hub. From my point of view this would make more sense (or would be more intuitive), because the concrete sender is named and so I would expect that this metric is related directly to the process of sending messages from the module to the edgeHub; but it is more astonishing to recognize that these numbers related to (maybe even long-time queued) messages from the module that got send from the edgeHub to the receiving module; also in a 1:1 relationship between sender and receiver module these numbers never differ, so they feel somehow redundant. Since the tree also shows a queue length I would have expected that it covers inbound and outbound messages - as explained before.

I know that I haven't added much new information, but I think basically I stated my assumptions/exceptions in the cited 
Regarding point 2 and 4: I think its fine to be EdgeHub-centric with the metrics - however it needs to be clear to the user what metric relates to inbound/outbound traffic.

Regarding the general question, what data and representation we wish for, I would like to just quote my previous comment.

> [...] I now have all the information at hand to analyze our scenarios - so many thanks for that. However it would be nice if this third graph could somehow depict if the sent messages are live or fed by the queue (e.g. by a different color) - maybe not in the sense of the single message (would violate the fifo character of the queue), but in a way that enables us to easily recognize the amount of messages that drain the backlog. So maybe it is even possible to combine the 2nd and 3rd graph and show send and received values as overlapping charts.

It's all about analyzing backlog scenarios; 
a) to be able to easily spot the root cause - e.g. just a temporary spike of incoming messages, or did the receiver decrease performance? 
b) recognize the capability to process the backlog - therefore the idea to give messages that arithmetically drain the backlog another color or combine the graphs that show inbound/outbound messages.

I know I haven't added much new information, but I still believe the assumptions/expectations from https://github.com/Azure/iotedge/issues/5662#issuecomment-988688754 give a starting point. For more specific questions, don't hesitate to contact me. And again thanks for your support.

**Comment by micahl** [comment]
@alaendle really appreciate all the feedback.  I've made more adjustments to the workbooks based on your input.  Try it out via [https://portal.azure.com/`?feature.workbookGalleryBranch=users%2Fmicahl%2Fux-polish`](https://portal.azure.com/?feature.workbookGalleryBranch=users%2Fmicahl%2Fux-polish)

**Comment by alaendle** [comment]
@micahl - Thanks for the update, it's really getting better and better (at least in my opinion, which of course is strongly influenced by our way of using the IoT edge device). However I could not explain the value which is presented as `Rx` - if I look at the last hour of data it seems unreasonable high (at least if it really only considers a one hour time slot). But maybe my expectations are just wrong again? However by looking at other durations (e.g. even 14 days) I still have a strong deviation between `Rx` and `Tx`.
![grafik](https://user-images.githubusercontent.com/969523/149713917-6a6e81ad-869f-4c92-ac54-4eaa93abda6e.png)

**Comment by micahl** [comment]
@alaendle right, in most cases it should be similar to the "Received (sum)"  shown in the "Messages from" graph on the lower right.  Looks like a potential error in  the query that I'll need to investigate.

**Comment by micahl** [comment]
Updated the query and believe it should be fixed now.

**Comment by alaendle** [comment]
Thanks numbers for `Tx`and `Rx` look much better now. Another area for improvement might be the chart presenting the free disk space.
![grafik](https://user-images.githubusercontent.com/969523/150081617-e92c0b6a-2ad7-40a5-a9f7-a9094a1a6002.png)
First I'm unsure if a stacked graph is a good representation, because normally not the sum of the free space is the problem, but one specific volume. The other thing, since not all volumes are shown as numeric values, there is a good chance to overlook small and therefore problematic volumes (because they get invisible in the chart and also may bot be represented numerically in the legend). So maybe it's a good idea to create another chart that just represents the status quo of free storage (in percent) like you usually find it on grafana boards? Just to give an idea...
![grafik](https://user-images.githubusercontent.com/969523/150083005-166a52c0-e786-4253-bd1e-7004c840c1af.png)

**Comment by alaendle** [comment]
@micahl Think I have found one additional problem the might only occur if messages are queued - please take a look at this video. First the messages send by `visu-poc-rabbitmq-gateway::unide_process` are shown. But after diving into a a specific time span, the complete route is missing.

![2022-01-19_09h00_21](https://user-images.githubusercontent.com/969523/150096952-aefd2469-5def-407b-b9c9-07731bb18a01.gif)

So my guess is that the query that creates the tree withhold some results in such cases!?

**Comment by micahl** [comment]
@alaendle, hmm, it may be getting dropped because of the order of joins happening in the query.  E.g. if no metric for "messages sent / delivered" was emitted by that module in that timeframe

Maybe do a quick check by running this query in log analytics (adjusting the start / end time).
```
let startTime = ago(6d);
let endTime = ago(3d);
InsightsMetrics
    | where TimeGenerated between (startTime .. endTime)
    | where Name == "edgehub_messages_sent_total"
    | extend dimensions=parse_json(Tags)
    | where tostring(dimensions.edge_device) == "visu-poc-rabbitmq-gateway"
    | count;
```

The above might be best for us to track as a separate github issue.  I'm planning to close this github issue with the proposed changes to the workbooks.

**Comment by alaendle** [comment]
@micahl You are right; I will create a new issue for the dropped route; maybe you could still consider my feedback regarding the free disk space representation. Thanks for your great support!

**Comment by micahl** [comment]
@alaendle thank you for all the feedback!  My latest changes in the PR referenced above alter the disk space to show the % used over time.  The fleet view and health snapshot were both showing % used so it made sense to be consistent in the details view.

**Comment by alaendle** [comment]
@micahl - Looks very strange; as far as I interpret it the graph now shows the percentage of free space per volume in ratio to the total free space (otherwise I couldn't imagine why things should add up to 100%). Sorry to say that, but I don't know how this should be helpful - and please apologize if my previous comment wasn't clear enough. I would expect something that shows the percentage of free space per volume (not stacked and not adding up to 100%).
![grafik](https://user-images.githubusercontent.com/969523/150322502-9c5fcf0e-e50f-4589-bdb3-a7ccc0e0ef31.png)

**Comment by micahl** [comment]
@alaendle oops, meant to use a different chart that doesn't stack the values.  Updated.

**Comment by micahl** [comment]
Closing this issue. Updates and fixes to the workbook templates should be live in ~48hrs.
